util/merkle/src/triedbmut.rs

```diff
@@ -0,0 +1,539 @@
+
+                            let mut node_rlp = RlpNode::encoded(Some(o_branch), None);
+                            let b_hash = self.db.insert(&node_rlp[..]);
+
+                            n_child[n_partial.at(0) as usize] = Some(b_hash);
+                            n_child[n_path.at(0) as usize] =
+                                Some(self.insert_aux(n_path.mid(1), v_, n_child[n_path.at(0) as usize], old_val)?);
+
+                            node_rlp = RlpNode::encoded(Some(RlpNode::Branch(partial, n_child)), Some(common));
+                            let hash = self.db.insert(&node_rlp[..]);
+
+                            self.db.remove(&cur_node_hash.expect("Can not remove branch and make new branch"));
+
+                            Ok(hash)
+                        } else {
+                            // Insert leaf into the branch node
+                            let n_path = path.mid(common);
+
+                            key[n_path.at(0) as usize] =
+                                Some(self.insert_aux(n_path.mid(1), v_, key[n_path.at(0) as usize], old_val)?);
+
+                            let n_branch = RlpNode::Branch(partial, key);
+                            let node_rlp = RlpNode::encoded(Some(n_branch), None);
+                            let hash = self.db.insert(&node_rlp[..]);
+                            self.db.remove(&cur_node_hash.expect("Can not insert leaf into branch"));
+
+                            Ok(hash)
+                        }
+                    }
+                    None => {
+                        let node = RlpNode::Leaf(path, v_);
+                        let node_rlp = RlpNode::encoded(Some(node), None);
+                        let hash = self.db.insert(&node_rlp[..]);
+
+                        self.db.remove(&cur_node_hash.expect("Can not remove root and add new leaf"));
+
+                        Ok(hash)
+                    }
+                }
+            }
+            None => {
+                let node = RlpNode::Leaf(path, v_);
+                let node_rlp = RlpNode::encoded(Some(node), None);
+                let hash = self.db.insert(&node_rlp[..]);
+
+                Ok(hash)
+            }
+        }
+    }
+
+    /// Get auxiliary
+    fn get_aux(&self, path: NibbleSlice, cur_node_hash: Option<H256>) -> super::Result<Option<DBValue>> {
+        match cur_node_hash {
+            Some(hash) => {
+                let node_rlp = self.db.get(&hash).ok_or_else(|| Box::new(TrieError::IncompleteDatabase(hash)))?;
+
+                match RlpNode::decoded(&node_rlp[..]) {
+                    Some(RlpNode::Leaf(partial, v)) => {
+                        if partial == path {
+                            Ok(Some(v))
+                        } else {
+                            Ok(None)
+                        }
+                    }
+                    Some(RlpNode::Branch(partial, key)) => {
+                        if path.starts_with(&partial) {
```

Comment:
Is this check necessary?
It seems quite inefficient to run this check on every tree traversal....